Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Basic.lean

```diff
@@ -329,4 +357,167 @@ def stalkIso (x : Proj 𝒜) :
 
 end stalk
 
+noncomputable section ofGlobalSection
+
+open Limits
+
+variable {X : Scheme.{u}} (f : A →+* Γ(X, ⊤)) {x x' : Γ(X, ⊤)} {t t' : A} {d d' : ℕ}
+
+/-- Given a graded ring `A` and a map `f : A →+* Γ(X, ⊤)`,
+for each homogeneous `t` of positive degree, it induces a map from `D(f(t)) ⟶ D₊(t)`. -/
+def toBasicOpenOfGlobalSections (H : f t = x) (h0d : 0 < d) (hd : t ∈ 𝒜 d) :
+    (X.basicOpen x).toScheme ⟶ basicOpen 𝒜 t := by
+  refine ?_ ≫ (basicOpenIsoSpec _ _ hd h0d).inv
+  refine (X.isoOfEq (X.toSpecΓ_preimage_basicOpen x)).inv ≫ X.toSpecΓ ∣_ _ ≫ ?_
+  refine (basicOpenIsoSpecAway _).hom ≫
+    Spec.map (CommRingCat.ofHom (RingHom.comp ?_ (algebraMap _ (Localization.Away t))))
+  refine IsLocalization.map (M := .powers t) (T := .powers x) _ f ?_
+  · rw [← Submonoid.map_le_iff_le_comap, Submonoid.map_powers]
+    simp [H]
+
+@[reassoc]
+lemma homOfLE_toBasicOpenOfGlobalSections_ι
+    {H : f t = x} {h0d : 0 < d} {hd : t ∈ 𝒜 d} {H' : f t' = x'} {h0d' : 0 < d'} {hd' : t' ∈ 𝒜 d'}
+    {s : A} (hts : t * s = t') {n : ℕ} (hn : d + n = d') (hs : s ∈ 𝒜 n) :
+    X.homOfLE (by aesop) ≫ toBasicOpenOfGlobalSections 𝒜 f H h0d hd ≫ (basicOpen 𝒜 t).ι =
+      toBasicOpenOfGlobalSections 𝒜 f H' h0d' hd' ≫ (basicOpen 𝒜 t').ι := by
+  simp only [toBasicOpenOfGlobalSections, Scheme.isoOfEq_inv,
+    ← Scheme.Hom.resLE_eq_morphismRestrict, CommRingCat.ofHom_comp, Spec.map_comp,
+    Scheme.Hom.map_resLE_assoc, Category.assoc, basicOpenIsoSpec_inv_ι]
+  have hx'x : PrimeSpectrum.basicOpen x' ≤ PrimeSpectrum.basicOpen x := by
+    aesop (add simp PrimeSpectrum.basicOpen_mul)
+  rw [← Scheme.Hom.resLE_map_assoc _ (by simp [X.toSpecΓ_preimage_basicOpen]) hx'x]
+  congr 1
+  simp only [← Iso.inv_comp_eq]
+  subst hts hn
+  rw [← SpecMap_awayMap_awayι (𝒜 := 𝒜) hd h0d
+    hs rfl, basicOpenIsoSpecAway_inv_homOfLE_assoc (R := Γ(X, ⊤)) x (f s) x' (by simp [← H', H]),
+    Iso.inv_hom_id_assoc]
+  simp only [← Category.assoc, ← Spec.map_comp, ← CommRingCat.ofHom_comp]
+  congr 3
+  ext
+  simp only [Fin.isValue, RingHom.coe_comp, Function.comp_apply,
+    HomogeneousLocalization.algebraMap_apply, HomogeneousLocalization.val_awayMap]
+  simp only [← RingHom.comp_apply]
+  congr 1
+  apply IsLocalization.ringHom_ext (M := .powers t)
+  ext i
+  simp [IsLocalization.Away.awayToAwayRight_eq]
+
+variable (f : A →+* Γ(X, ⊤)) (hf : (HomogeneousIdeal.irrelevant 𝒜).toIdeal.map f = ⊤)
```

Comment:
Unless I am mistaken, the assumption `hf` could eventually be replaced by a formulation that is "local" on `X` (maybe using `IdealSheafData` instead of ideals in the ring of global sections). (Of course, this is equivalent when `X` is affine.) If this makes sense, would it be difficult to generalize the definition of the corresponding open cover?

Reply:
I think the "right definition" is have a line bundle and have `f` to map into the global sections of that but we are probably quite far from that. And the intention of this API is that this is only used for affine things and we glue them on the source via `OpenCover.glueMorphisms` by hand for now. I will think about your suggestion, whether there is a middle ground that is both possible in terms of what we have in mathlib and is stronger than what I have here.

Reply:
But I am quite certain that `hf` is equivalent to "`{ D(f(r)) | r in A homogeneous of positive degree }` covers `X` so we probably can't generalize `hf` without tweaking what `f` is?

Reply:
The way to "cheat" to avoid invertible sheaves is to pick an open cover `D+(rᵢ)` of `Proj A`, assign to each `rᵢ` an `sᵢ ∈ Γ(X, Uᵢ)` satisfying some co-cycle condition and then you can state some local condition on `A` that is equivalent to the invertible sheaf to be base point free. But I think this is too far removed from this API so I think this API is still worth adding.